src/main/java/com/pingcap/tikv/codec/DecimalUtils.java

```diff
@@ -30,18 +30,26 @@
      * @param cdi cdi is source data.
      * */
     public static double readDecimalFully(CodecDataInput cdi) {
-        if (cdi.size() < 3) {
+        if (cdi.available() < 3) {
             throw new IllegalArgumentException("insufficient bytes to read value");
         }
+
+        MyDecimal dec = new MyDecimal();
+        List<Integer> data = new ArrayList<>();
```

Comment:
AutoBoxing is known as very heavy. I got this point. I was to rush to jump to the conclusion and will do it right now.  You are absolutely right. `Ints.toArray`could reduce  the burden. Another approbate approach is needed here. Thanks for the very detail investigation.